File: prototype/rails/src/app/actions/escrow/run_weekly_payout.rb
module Escrow
  class RunWeeklyPayout
    def call(as_of:)
      period = Domain::Escrow::PayoutPeriod.ending_before(as_of)

      ActiveRecord::Base.transaction do
        payable_balances(period).map { |seller_id, balance| pay_out(seller_id, balance.available, period, as_of) }
      end
    end

    private

    def payable_balances(period)
      LedgerEntry.occurred_by(period.ends_at)
                 .order(:seller_id, :id)
                 .group_by(&:seller_id)
                 .transform_values { |entries| Domain::Escrow::LedgerBalance.from(entries.map(&:to_movement)) }
                 .select { |_seller_id, balance| balance.payable? }
    end

    def pay_out(seller_id, available, period, as_of)
      payout = Payout.create!(
        seller_id: seller_id,
        period_start: period.first_day,
        period_end: period.last_day,
        amount_cents: available.cents,
        paid_at: as_of
      )
      movement = Domain::Escrow::LedgerMovement.payout(available)

      # Dated at the close of the period it settles, so a second run of the same
      # period already counts the money as sent and pays nothing.
      payout.ledger_entries.create!(
        seller_id: seller_id,
        entry_type: movement.entry_type,
        amount_cents: movement.amount.cents,
        occurred_at: period.ends_at
      )

      payout
    end
  end
end

File: prototype/php/src/app/Actions/Escrow/RunWeeklyPayout.php
<?php

declare(strict_types=1);

namespace App\Actions\Escrow;

use App\Domain\Escrow\LedgerBalance;
use App\Domain\Escrow\LedgerMovement;
use App\Domain\Escrow\PayoutPeriod;
use App\Domain\Money\Money;
use App\Models\LedgerEntry;
use App\Models\Payout;
use DateTimeImmutable;
use Illuminate\Support\Collection;
use Illuminate\Support\Facades\DB;

final readonly class RunWeeklyPayout
{
    /**
     * @return list<Payout>
     */
    public function __invoke(DateTimeImmutable $asOf): array
    {
        $period = PayoutPeriod::endingBefore($asOf);

        return DB::transaction(fn (): array => array_values($this->balancesBySeller($period)
            ->filter(fn (LedgerBalance $balance): bool => $balance->isPayable())
            ->map(fn (LedgerBalance $balance, string $sellerId): Payout => $this->payOut($sellerId, $balance->available, $period, $asOf))
            ->all()));
    }

    /**
     * @return Collection<string, LedgerBalance>
     */
    private function balancesBySeller(PayoutPeriod $period): Collection
    {
        return LedgerEntry::query()
            ->occurredBy($period->end)
            ->totalledByType()
            ->get()
            ->groupBy('seller_id')
            ->map(fn (Collection $entries): LedgerBalance => LedgerBalance::from(
                array_values($entries->map(fn (LedgerEntry $entry): LedgerMovement => $entry->toMovement())->all()),
            ));
    }

    private function payOut(string $sellerId, Money $available, PayoutPeriod $period, DateTimeImmutable $asOf): Payout
    {
        $payout = Payout::create([
            'seller_id' => $sellerId,
            'period_start' => $period->start,
            'period_end' => $period->end,
            'amount_cents' => $available->cents,
            'paid_at' => $asOf,
        ]);

        $movement = LedgerMovement::payout($available);

        LedgerEntry::create([
            'seller_id' => $sellerId,
            'payout_id' => $payout->id,
            'type' => $movement->type,
            'amount_cents' => $movement->amount->cents,
            // Dated inside the period it settles so a re-run of the same period
            // sees the money as already paid.
            'occurred_at' => $period->end,
        ]);

        return $payout;
    }
}
